server/src/internalClusterTest/java/org/elasticsearch/action/bulk/WriteMemoryLimitsIT.java

```diff
@@ -0,0 +1,203 @@
+        assertAcked(prepareCreate(index, Settings.builder()
+            .put(IndexMetadata.SETTING_NUMBER_OF_SHARDS, 1)
+            .put(IndexMetadata.SETTING_NUMBER_OF_REPLICAS, 1)));
+        ensureGreen(index);
+
+        IndicesStatsResponse response = client().admin().indices().prepareStats(index).get();
+        String primaryId = Stream.of(response.getShards())
+            .map(ShardStats::getShardRouting)
+            .filter(ShardRouting::primary)
+            .findAny()
+            .get()
+            .currentNodeId();
+        String replicaId = Stream.of(response.getShards())
+            .map(ShardStats::getShardRouting)
+            .filter(sr -> sr.primary() == false)
+            .findAny()
+            .get()
+            .currentNodeId();
+        String primaryName = client().admin().cluster().prepareState().get().getState().nodes().get(primaryId).getName();
+        String replicaName = client().admin().cluster().prepareState().get().getState().nodes().get(replicaId).getName();
+
+        final CountDownLatch replicationSendPointReached = new CountDownLatch(1);
+        final CountDownLatch latchBlockingReplicationSend = new CountDownLatch(1);
+        final CountDownLatch newActionsSendPointReached = new CountDownLatch(2);
+        final CountDownLatch latchBlockingReplication = new CountDownLatch(1);
+
+        TransportService primaryService = internalCluster().getInstance(TransportService.class, primaryName);
+        final MockTransportService primaryTransportService = (MockTransportService) primaryService;
+        TransportService replicaService = internalCluster().getInstance(TransportService.class, replicaName);
+        final MockTransportService replicaTransportService = (MockTransportService) replicaService;
+
+        primaryTransportService.addSendBehavior((connection, requestId, action, request, options) -> {
+            if (action.equals(TransportShardBulkAction.ACTION_NAME + "[r]")) {
+                try {
+                    replicationSendPointReached.countDown();
+                    latchBlockingReplicationSend.await();
+                } catch (InterruptedException e) {
+                    throw new IllegalStateException(e);
+                }
+            }
+            connection.sendRequest(requestId, action, request, options);
+        });
+
+        final BulkRequest bulkRequest = new BulkRequest();
+        int totalRequestSize = 0;
+        for (int i = 0; i < 80; ++i) {
+            IndexRequest request = new IndexRequest(index).id(UUIDs.base64UUID())
+                .source(Collections.singletonMap("key", randomAlphaOfLength(50)));
+            totalRequestSize += request.ramBytesUsed();
+            bulkRequest.add(request);
+        }
+
+        final long bulkRequestSize = bulkRequest.ramBytesUsed() + WriteMemoryLimits.WRITE_REQUEST_BYTES_OVERHEAD;
+        final long bulkShardRequestSize = totalRequestSize + WriteMemoryLimits.WRITE_REQUEST_BYTES_OVERHEAD;
+
```

Comment:
Maybe add a couple of assertions to demonstrate that we do get some byte-size out like `bulkRequestSize > totalRequestSize` and `totalRequestSize > 80 * 4` or similar? I think all assertions here would otherwise succeed if `ramBytesUsed()` returns 0.